### ISSUE TITLE
CSP is not removed for exception pages in 6.1.4

### ISSUE THREAD

**Issue by iquito** [symfony/symfony#47405]
### Symfony version(s) affected

6.1.4

### Description

When upgrading from 6.1.3 to 6.1.4 the exception pages do not remove the Content-Security-Policy anymore, which leads to all styles/scripts failing to be executed if a restrictive CSP is set.

### How to reproduce

Setting a Content-Security-Policy on like:

```
Content-Security-Policy: script-src 'self';  style-src 'self'; img-src 'self' data:
```

And then open a page in a Symfony project in development/debug mode where an exception is thrown.

### Possible Solution

As far as I can tell, the CSP should be removed in the HttpKernel ErrorListener (https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/HttpKernel/EventListener/ErrorListener.php) in `removeCspHeader`, which is a response event listener. In my project with 6.1.4 `removeCspHeader` is called before `onKernelException`, yet `removeCspHeader` relies on an attribute being set in `onKernelException` to actually remove the CSP header, so the order in which these are called seems to be wrong with 6.1.4.

After more investigation, I found out that in 6.1.3 the `removeCspHeader` listener is called twice (again at the very end) and in 6.1.4 it is only called once, so the order has not changed, but one execution is now not happening. See my next comment for the comparisons of event listener execution.

**Comment by iquito** [comment]
After testing with 6.1.3 and 6.1.4 these are the orders of event listener executions in 6.1.3:

- My own CSP response listener with priority 10
- `ErrorListener::removeCspHeader`
- `ErrorListener::onKernelException`
- My own CSP response listener with priority 10
- `ErrorListener::removeCspHeader`

This is the execution in 6.1.4:

- My own CSP response listener with priority 10
- `ErrorListener::removeCspHeader`
- `ErrorListener::onKernelException`
- My own CSP response listener with priority 10

So with 6.1.4 it seems the difference is that `removeCspHeader` is not called a second time.

**Comment by xabbuh** [comment]
Can you create a small example application that allows to reproduce your issue?

**Comment by iquito** [comment]
After some more investigation I found out where the problem occurs: in `RequestDataCollector` on line 113:

```php
'stateless_check' => $this->requestStack?->getMainRequest()->attributes->get('_stateless') ?? false,
```

`RequestDataCollector` is called by `ProfilerListener`, and `getMainRequest` on that line can return null, which happens in my application, leading in my application to:

```
Warning: Attempt to read property "attributes" on null
```

Which is elevated to an exception in development, leading to all other response event listeners not being called anymore. Guarding against null with `$this->requestStack?->getMainRequest()?->attributes->get('_stateless')` removes the warning.

**Comment by stof** [comment]
`getMainRequest` returning `null` at the time we profile might indicate a bug in 6.1.4. We would need to investigate how https://github.com/symfony/symfony/pull/47358 affected this case (we might have missed something in the state changes of the RequestStack).

However, once we find the root cause of this issue, I think it would *also* make sense for RequestDataCollector to handle the `null` value gracefully.

**Comment by iquito** [comment]
I was thinking the same thing, as even after an exception occured `getMainRequest` should still return a request as far as I would understand it, just like it did before the exception occured.

**Comment by xabbuh** [comment]
@iquito If you can manage to create a reproducer application, I will happily try to find the root cause for this issue.

**Comment by iquito** [comment]
I found the last piece of information explaining this bug and have created a new symfony project from scratch showcasing the bug: [exceptionpagebug.zip](https://github.com/symfony/symfony/files/9482383/exceptionpagebug.zip)

In my case, a special exception was thrown by the application (`SpecialException` in the demo app) and handled in in an event listener by the application to set a new valid response there. During the generation of that response, there was an unexpected exception which was not handled, leading to the Symfony error page.

I believe this code part in `HttpKernel::handle`, which was changed in 6.1.4, is the cause of this new behavior:

```php
$this->requestStack->push($request);
try {
    return $this->handleRaw($request, $type);
} catch (\Exception $e) {
    if ($e instanceof RequestExceptionInterface) {
        $e = new BadRequestHttpException($e->getMessage(), $e);
    }
    if (false === $catch) {
        $this->finishRequest($request, $type);

        throw $e;
    }

    return $this->handleThrowable($e, $request, $type);
} finally {
    $this->requestStack->pop();
}
```

Using `$this->requestStack->push` and `$this->requestStack->pop` is new in 6.1.4, but the behavior change comes from using `finally`: even if there is another exception in the catch block the initial request is removed by the finally block, which is one of the surprising side effects of using `finally` - it is always executed. By the time the profiler wants to know about the main request, it has been removed by that finally block, even though the rest of the control flow was interrupted. The following code without finally behaves more like it did before:

```php
$this->requestStack->push($request);
try {
    $response = $this->handleRaw($request, $type);
} catch (\Exception $e) {
    if ($e instanceof RequestExceptionInterface) {
        $e = new BadRequestHttpException($e->getMessage(), $e);
    }
    if (false === $catch) {
        $this->finishRequest($request, $type);

        throw $e;
    }

    $response = $this->handleThrowable($e, $request, $type);
}

$this->requestStack->pop();

return $response;
```

In this case, if an exception is thrown in the catch block, `$this->requestStack->pop()` is never reached, which makes the main request still available to the profiler.

**Comment by wouterj** [comment]
@iquito do you have time to open a PR with this change (and a test case proofing the bug)?

What you're describing makes sense to me, and we've found this behavior in another profiler-related bug as well (#47577).

**Comment by gnat42** [comment]
https://github.com/symfony/symfony/commit/e35d6bab3e1a634ce0bd73a2c678d5510a8a76cd is the commit that changed the behaviour. What I don't understand about the change is *why*. What bug/problem was being fixed? Knowing this would allow for a solution that doesn't break that change but allows for handling errors in the handleThrowable method

**Comment by wouterj** [comment]
@gnat42 that commit is part of this PR, which gives a bit more context: https://github.com/symfony/symfony/pull/47358 (it also added tests, so in theory if those keep green we don't break the original behavior that was fixed).

**Comment by gnat42** [comment]
Ok, so using Xdebug, it seems that the finally {} block is executed *before* the catch, which flies in the face of the php docs. I've tried to create a reproducing situation with a simple php script but it behaves as expected in the documentation. I think i may be because the handle method is somewhat used in a recursive fashion. I'll continue to investigate

**Comment by nicolas-grekas** [comment]
Should be fixed by #47857
